File: python_agents/workflow_engine.py
"""Python-based LangChain workflow engine"""

import asyncio
import uuid
from datetime import datetime
from typing import Dict, Any, Optional, Callable
from workflow_state import WorkflowState, WorkflowResponse, AgentProgress
from agents import jira_analyst_agent, technical_architect_agent, product_manager_agent
from vector_store import vector_store


class PythonWorkflowEngine:
    """Python-based workflow engine for LangChain agents"""
    
    def __init__(self):
        self.active_workflows: Dict[str, WorkflowState] = {}
        self.progress_callbacks: Dict[str, Callable] = {}
        
        # Load sample documents into vector store
        vector_store.load_sample_documents()
        print(f"[Python] Vector store initialized with {vector_store.get_document_count()} documents")
    
    async def start_workflow(self, jira_ticket_id: str, progress_callback: Optional[Callable] = None) -> str:
        """Start a new LangChain workflow"""
        workflow_id = str(uuid.uuid4())
        
        print(f"[Python] Starting LangChain workflow for ticket: {jira_ticket_id}")
        
        # Initialize workflow state
        state: WorkflowState = {
            'jira_ticket_id': jira_ticket_id,
            'current_step': 'jira_analyst',
            'currentAgent': 'jira-analyst',  # Set initial agent
            'metadata': {
                'start_time': datetime.now().isoformat(),
                'engine_type': 'langchain-python',
                'version': '2.0'
            }
        }
        
        self.active_workflows[workflow_id] = state
        
        if progress_callback:
            self.progress_callbacks[workflow_id] = progress_callback
        
        # Start execution asynchronously
        asyncio.create_task(self._execute_workflow(workflow_id))
        
        return workflow_id
    
    async def _execute_workflow(self, workflow_id: str):
        """Execute the complete workflow"""
        state = self.active_workflows.get(workflow_id)
        if not state:
            print(f"[Python] Workflow {workflow_id} not found")
            return
        
        progress_callback = self.progress_callbacks.get(workflow_id)
        
        try:
            # Step 1: Jira Analyst Agent
            print(f"[Python] Executing JiraAnalystAgent for workflow {workflow_id}")
            
            async def analyst_progress(agent: str, message: str, progress: int):
                if progress_callback:
                    await progress_callback(AgentProgress(
                        workflow_id=workflow_id,
                        agent=agent,
                        message=message,
                        progress=progress
                    ))
            
            analyst_result = await jira_analyst_agent.execute(state, analyst_progress)
            
            if analyst_result.get('error'):
                raise Exception(analyst_result['error'])
            
            # Update state with analyst results
            state.update(analyst_result)
            state['currentAgent'] = 'completed'  # Mark as completed
            
            if progress_callback:
                await progress_callback(AgentProgress(
                    workflow_id=workflow_id,
                    agent='jira-analyst',
                    message='Ticket data fetch and Excel export completed',
                    progress=100,
                    completed=True
                ))
            
            # Skip other agents - only do data fetch and Excel export
            
            # Workflow completed successfully
            state['current_step'] = 'completed'
            state['currentAgent'] = 'completed'  # Mark as completed
            
            if progress_callback:
                await progress_callback(AgentProgress(
                    workflow_id=workflow_id,
                    agent='workflow',
                    message='LangChain workflow completed successfully',
                    progress=100,
                    completed=True
                ))
            
            print(f"[Python] Workflow {workflow_id} completed successfully")
            
        except Exception as error:
            print(f"[Python] Workflow {workflow_id} failed: {error}")
            state['error'] = str(error)
            state['current_step'] = 'failed'
            
            if progress_callback:
                await progress_callback(AgentProgress(
                    workflow_id=workflow_id,
                    agent='workflow',
                    message=f'Workflow failed: {str(error)}',
                    progress=100,
                    completed=True
                ))
    
    def get_workflow_status(self, workflow_id: str) -> Optional[WorkflowState]:
        """Get current workflow status"""
        return self.active_workflows.get(workflow_id)
    
    def get_all_workflows(self) -> Dict[str, WorkflowState]:
        """Get all active workflows"""
        return self.active_workflows.copy()
    
    async def cleanup_workflow(self, workflow_id: str):
        """Clean up completed workflow"""
        if workflow_id in self.active_workflows:
            del self.active_workflows[workflow_id]
        if workflow_id in self.progress_callbacks:
            del self.progress_callbacks[workflow_id]


# Global workflow engine instance
workflow_engine = PythonWorkflowEngine()

File: server/services/langchain/workflowEngine.ts
import { storage } from '../../storage.js';
import { 
  langChainJiraAnalystAgent, 
  langChainTechnicalArchitectAgent, 
  langChainProductManagerAgent 
} from './agents.js';
import { type WorkflowState, type AgentContext } from './workflowState.js';
import { type Workflow } from '@shared/schema';
import WebSocket from 'ws';

export class LangChainWorkflowEngine {
  private activeWorkflows: Map<string, WorkflowState> = new Map();
  private wsClients: Set<WebSocket> = new Set();

  addWebSocketClient(ws: WebSocket) {
    this.wsClients.add(ws);
    ws.on('close', () => {
      this.wsClients.delete(ws);
    });
  }

  private broadcast(message: any) {
    const messageStr = JSON.stringify(message);
    this.wsClients.forEach(ws => {
      if (ws.readyState === WebSocket.OPEN) {
        ws.send(messageStr);
      }
    });
  }

  async startWorkflow(jiraTicketId: string): Promise<string> {
    console.log(`[LangChain] Starting enhanced workflow for ticket: ${jiraTicketId}`);
    
    // Create workflow record
    const workflow = await storage.createWorkflow({
      jiraTicketId,
      status: 'pending',
      currentAgent: 'jira-analyst',
    });

    // Initialize enhanced workflow state
    const state: WorkflowState = {
      jira_ticket_id: jiraTicketId,
      current_step: 'jira_analyst',
      metadata: {
        startTime: new Date().toISOString(),
        engineType: 'langchain',
        version: '2.0'
      }
    };
    
    this.activeWorkflows.set(workflow.id, state);
    
    // Broadcast workflow started
    this.broadcast({
      type: 'langchain_workflow_started',
      workflowId: workflow.id,
      jiraTicketId,
      engineType: 'langchain',
    });

    // Start execution asynchronously
    this.executeWorkflow(workflow.id).catch(error => {
      console.error(`[LangChain] Workflow ${workflow.id} failed:`, error);
    });

    return workflow.id;
  }

  private async executeWorkflow(workflowId: string): Promise<void> {
    const state = this.activeWorkflows.get(workflowId);
    if (!state) {
      throw new Error(`[LangChain] Workflow ${workflowId} not found`);
    }

    const context: AgentContext = {
      workflowId,
      broadcast: this.broadcast.bind(this),
      updateStorage: async (updates: Partial<WorkflowState>) => {
        Object.assign(state, updates);
        await storage.updateWorkflow(workflowId, {
          jiraTicketData: state.jira_ticket_data,
          ragContext: state.rag_context,
          impactAnalysis: state.impact_analysis,
          solutionArchitecture: state.solution_architecture,
          prd: state.prd,
          currentAgent: state.current_step,
        });
      }
    };

    try {
      // Update workflow status
      await storage.updateWorkflow(workflowId, { 
        status: 'running',
        currentAgent: 'jira-analyst' 
      });

      this.broadcast({
        type: 'langchain_workflow_updated',
        workflowId,
        status: 'running',
        currentAgent: 'jira-analyst',
      });

      // Step 1: Enhanced Jira Analyst Agent
      console.log(`[LangChain] Executing JiraAnalystAgent for workflow ${workflowId}`);
      const analystResult = await langChainJiraAnalystAgent.execute(state, context);
      
      if (analystResult.error) {
        throw new Error(analystResult.error);
      }

      Object.assign(state, analystResult);
      await context.updateStorage(state);
      
      await storage.createActivity({
        workflowId,
        type: 'langchain_ticket_analyzed',
        message: `[LangChain] ${state.jira_ticket_id} analyzed with enhanced RAG context`,
      });

      // Update metrics
      const metrics = await storage.getMetrics();
      if (metrics) {
        await storage.updateMetrics({
          ticketsProcessed: (metrics.ticketsProcessed || 0) + 1,
        });
      }

      this.broadcast({
        type: 'langchain_agent_completed',
        workflowId,
        agent: 'jira-analyst',
        message: 'Enhanced ticket analysis with semantic RAG retrieval completed',
      });

      // Step 2: Enhanced Technical Architect Agent
      console.log(`[LangChain] Executing TechnicalArchitectAgent for workflow ${workflowId}`);
      const architectResult = await langChainTechnicalArchitectAgent.execute(state, context);
      
      if (architectResult.error) {
        throw new Error(architectResult.error);
      }

      Object.assign(state, architectResult);
      await context.updateStorage(state);
      
      await storage.createActivity({
        workflowId,
        type: 'langchain_impact_completed',
        message: `[LangChain] Chain-of-Thought impact analysis and solution architecture completed`,
      });

      // Update metrics
      if (metrics) {
        await storage.updateMetrics({
          impactAnalyses: (metrics.impactAnalyses || 0) + 1,
          solutionArchitectures: (metrics.solutionArchitectures || 0) + 1,
        });
      }

      this.broadcast({
        type: 'langchain_agent_completed',
        workflowId,
        agent: 'tech-architect',
        message: 'Chain-of-Thought reasoning for impact analysis and solution design completed',
      });

      // Step 3: Enhanced Product Manager Agent
      console.log(`[LangChain] Executing ProductManagerAgent for workflow ${workflowId}`);
      const pmResult = await langChainProductManagerAgent.execute(state, context);
      
      if (pmResult.error) {
        throw new Error(pmResult.error);
      }

      Object.assign(state, pmResult);
      await context.updateStorage(state);
      
      await storage.updateWorkflow(workflowId, {
        status: 'completed',
        currentAgent: null,
        prd: state.prd,
      });

      await storage.createActivity({
        workflowId,
        type: 'langchain_prd_completed',
        message: `[LangChain] Structured PRD document generated with Pydantic validation`,
      });

      // Update metrics
      if (metrics) {
        await storage.updateMetrics({
          prdsGenerated: (metrics.prdsGenerated || 0) + 1,
        });
      }

      this.broadcast({
        type: 'langchain_workflow_completed',
        workflowId,
        message: 'Enhanced LangChain workflow completed with structured outputs',
        engineType: 'langchain',
      });

      console.log(`[LangChain] Workflow ${workflowId} completed successfully`);

    } catch (error) {
      console.error(`[LangChain] Workflow ${workflowId} failed:`, error);
      
      await storage.updateWorkflow(workflowId, {
        status: 'failed',
        error: error instanceof Error ? error.message : 'Unknown error',
      });

      this.broadcast({
        type: 'langchain_workflow_failed',
        workflowId,
        error: error instanceof Error ? error.message : 'Unknown error',
        engineType: 'langchain',
      });
    } finally {
      this.activeWorkflows.delete(workflowId);
    }
  }

  async getWorkflowStatus(workflowId: string): Promise<Workflow | undefined> {
    return storage.getWorkflow(workflowId);
  }

  async getAllWorkflows(): Promise<Workflow[]> {
    return storage.getAllWorkflows();
  }

  // Enhanced workflow diagnostics
  async getWorkflowDiagnostics(workflowId: string) {
    const workflow = await storage.getWorkflow(workflowId);
    const state = this.activeWorkflows.get(workflowId);
    
    return {
      workflow,
      state,
      isActive: this.activeWorkflows.has(workflowId),
      engineType: 'langchain',
      capabilities: [
        'Enhanced RAG with semantic search',
        'Chain-of-Thought reasoning',
        'Structured Pydantic outputs',
        'Real-time progress tracking'
      ]
    };
  }
}

export const langChainWorkflowEngine = new LangChainWorkflowEngine();